src/Dialect/ONNX/Rewrite.cpp

```diff
@@ -331,6 +331,41 @@ class BinaryOpBroadcastAxisPattern : public OpRewritePattern<OP_TYPE> {
   }
 };
 
+// =============================================================================
+// Rewrite pattern for CastLikeOp to CastOp (not handled in Rewrite.td).
+// =============================================================================
+// A pattern to turn
+//   `CastLikeOp(input, target_type) attribute -> saturate=1`
+// into
+//   `CastOp(input) attribute -> saturate=1`
+// we will force all CastLikeOps to become CastOp because
+// they are not optimized for constant propagation.
+class PropagateCastLikeToCastPattern : public OpRewritePattern<ONNXCastLikeOp> {
+public:
+  using OpRewritePattern<ONNXCastLikeOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(
+      ONNXCastLikeOp castLikeOp, PatternRewriter &rewriter) const override {
+    Operation *op = castLikeOp.getOperation();
+    Location loc = castLikeOp.getLoc();
+
+    Value input = op->getOperand(0);
+
+    // The output type will be the same as the target_type or the second input
+    Type outputType = castLikeOp.getTargetType()
+                          .getType()
+                          .cast<ShapedType>()
+                          .getElementType();
+
+    // Replace
+    MultiDialectBuilder<OnnxBuilder> create(rewriter, loc);
+    Value res;
+    res = create.onnx.cast(input, outputType);
+    rewriter.replaceOp(castLikeOp, res);
+    return success();
+  };
+};
+
```

Comment:
@hamptonm1 if you move this into Decompose pass, then we don't need its shape inference.

Reply:
@tungld Ahh thanks for that!! Let me try it out :)

Reply:
@tungld That did not work unfortunately.... I still see the error even when I move it to decompose pass because I still specify ONNXCastLike:

```
ld: Undefined symbols:
  mlir::ONNXCastLikeOp::inferShapes(std::__1::function<void (mlir::Region&)>), referenced from:
      mlir::detail::ShapeInferenceOpInterfaceInterfaceTraits::Model<mlir::ONNXCastLikeOp>::inferShapes(mlir::detail::ShapeInferenceOpInterfaceInterfaceTraits::Concept const*, mlir::Operation*, std::__1::function<void (mlir::Region&)>) in libOMONNXOps.a[5](ONNXDialect.cpp.o)
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Reply:
You should declare that this op will be illegal otherwise the rule is not applied. See https://github.com/onnx/onnx-mlir/blob/main/src/Transform/ONNX/Decompose.cpp#L949

Reply:
kk testing now

Reply:
@tungld Okay that worked but now the model does not work. I get the following error:

```
Onnx-mlir command: ./Debug/bin/onnx-mlir --O3 --EmitLib --onnx-op-stats TXT model.onnx -v
The ONNX model has 42970 elements in its initializers. This value would be close to and greater than the number of parameters in the model. Because there is no way to exactly count the number of parameters, this value can be used to have a rough idea of the number of parameters in the model.
loc("n1"): error: failed to legalize operation 'onnx.CastLike' that was explicitly marked illegal
```